### DISCUSSION TITLE
Record/play custom messages without having to source workspace

### DISCUSSION THREAD

**Discussion by weidinger-c** [foxglove/mcap#1089] [General] [answered]
As I understood, MCAP is, like ROS 1 Bags, an encapsulated container that contains all msg definitions.
Right now when recording MCAP bags in ROS 2 Humble, I still have to source a workspace that contains custom messages to be able to record them.

Same goes when I try to play an MCAP bag with "ros2 bag play". Custom messages are skipped with the message:
[WARN] [1689771449.920892226] [rosbag2_player]: Ignoring a topic '/TOPIC', reason: package 'PACKAGE' not found, searching: [/ros_ws/install/ros2bag_tools, /ros_ws/install/rosbag2_tools, /ros_ws/install/rosbag2_rviz_plugins, /opt/ros/humble].
This is because I played the ROS 2 Bag in an environment, where the custom messages were not installed.

Playing custom messages was possible with ROS 1 bags, is there anything planned to support this behaviour with ROS 2 and MCAP?

**Comment by emersonknapp** [comment, marked as answer]
As of current ROS 2 state, message types must be installed in the local space in order to play them back in the bag. The schemas stored in the MCAP file are for use by other tooling, not the ROS 2 system itself.

To highlight a few details about it, the `rclcpp::GenericPublisher` that is used by the Rosbag2 `Player` class still requires the ability to fetch the appropriate `typesupport` for the type. The benefit from the "generic" publisher is the ability to publish already-serialized messages, saving performance.

The ability to playback without message types installed locally doesn't yet exist in the ros2 core, so it's not an MCAP issue specifically. There is some work begun in `rclcpp` (https://github.com/ros2/ros2/issues/1374) to support "dynamic" subscription, which would require only a schema description locally and would lead naturally into the "dynamic" publisher. That feature is a part of making this "fully encapsulated playback" possible.

**Reply by emersonknapp** [reply]
Further note - we had https://github.com/ros2/rosbag2/issues/1373 filed for "dynamic recording", but I've just filed a related https://github.com/ros2/rosbag2/issues/1428 to track the "dynamic playback" complementary feature

**Reply by weidinger-c** [reply]
Thank you for that clarification. Looking forward to using "dynamic recording/playback“ at some point!
